File: DesignPatternPractice_CSharp/Source/AbstractFactoryPattern.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Reflection;

namespace DesignPatternPractice.Source
{
    internal class AbstractFactoryPattern
    {
        internal abstract class Animal
        {
            internal abstract void Speak();
        }

        internal abstract class Dog : Animal
        {
            internal abstract void PlayBall();
        }

        internal class DogXY : Dog
        {
            internal override void Speak() => Console.WriteLine("I am a XY dog");
            internal override void PlayBall() => Console.WriteLine("XY Dog plays ball");
        }

        internal class DogXX : Dog
        {
            internal override void Speak() => Console.WriteLine("I am a XX dog");
            internal override void PlayBall() => Console.WriteLine("XX Dog plays ball");
        }

        internal interface IAnimalFactory
        {
            public Dog CreateDog();
        }

        internal class AnimalXYFactory : IAnimalFactory
        {
            public Dog CreateDog() => new DogXY();
        }

        internal class AnimalXXFactory : IAnimalFactory
        {
            public Dog CreateDog() => new DogXX();
        }

        public static void Run()
        {
            IAnimalFactory dogXYFactory = new AnimalXYFactory();
            Animal dogXY = dogXYFactory.CreateDog();
            dogXY.Speak();
            (dogXY as DogXY)?.PlayBall();

            IAnimalFactory dogXXFactory = new AnimalXXFactory();
            Animal dogXX = dogXXFactory.CreateDog();
            dogXX.Speak();
            (dogXX as DogXX)?.PlayBall();
        }
    }
}

File: DesignPatternPractice_Python/AbstractFactoryPattern.py
from abc import ABCMeta, abstractmethod
from typing import Type


class AbstractFactoryPattern:
    @staticmethod
    def run():
        class Animal(metaclass=ABCMeta):
            @abstractmethod
            def speak(self):
                pass

        class Dog(Animal):
            pass

        class DogXY(Dog):
            def speak(self):
                print("I am a XY dog")

        class DogXX(Dog):
            def speak(self):
                print("I am a XX dog")

        class AnimalFactory(metaclass=ABCMeta):
            @abstractmethod
            def create_dog(self) -> Animal:
                pass

        class AnimalXYFactory(AnimalFactory):
            def create_dog(self) -> Animal:
                return DogXY()

        class AnimalXXFactory(AnimalFactory):
            def create_dog(self) -> Animal:
                return DogXX()

        dog_xy_factory = AnimalXYFactory()  # type: AnimalXYFactory
        dog_xy_factory.create_dog().speak()

        dog_xx_factory = AnimalXXFactory()  # type: AnimalXYFactory
        dog_xx_factory.create_dog().speak()
